### ISSUE TITLE
Container failing to start correctly with "kill usage" message

### ISSUE THREAD

**Issue by deosrc** [pi-hole/docker-pi-hole#874]
This is a: Bug

## Details

The Pi-Hole container does not start correctly with the following "kill usage" message appearing in the logs (surrounding messages included for context):

```text
 ::: Docker start setup complete
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
  Pi-hole version is v5.3.1 (Latest: v5.3.1)
  AdminLTE version is v5.5 (Latest: v5.5)
  FTL version is v5.8.1 (Latest: v5.8.1)
[cont-init.d] 20-start.sh: exited 0.
```

## Related Issues
- [x] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

#555 
#405 

The above issues seem to be general support threads with various generic fixes. I am including specific details here on my findings so figured a new issue would be best. If maintainers believe this would be best in one of the above threads instead, let me know and I'll move the content.

## How to reproduce the issue 

1. Environment data
  * Operating System: RaspberryPi OS 
  * Hardware: RasPi4B
  * Kernel Architecture: ArmV7
  * Docker Install Info and version: 
    - Software source: K3s
    - Supplimentary Software: Kubernetes Dashboard
  * Hardware architecture: ARMv7

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here

```yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: pihole
  namespace: home
  labels:
    app: pihole
spec:
  replicas: 1
  selector:
    matchLabels:
      app: pihole
  serviceName: pihole
  template:
    metadata:
      labels:
        app: pihole
    spec:
      restartPolicy: Always
      hostNetwork: true
      dnsPolicy: "None"
      dnsConfig:
        nameservers:
          - 127.0.0.1
          - 1.1.1.1
      containers:
        - name: pihole
          image: pihole/pihole:v5.8.1
          securityContext:
            capabilities:
              add:
                - NET_ADMIN
          ports:
            - containerPort: 31180
              name: http
            - containerPort: 53
              protocol: UDP
              name: dns-udp
            - containerPort: 53
              protocol: TCP
              name: dns-tcp
            - containerPort: 67
              protocol: UDP
              name: dhcp
          volumeMounts:
            - name: config
              mountPath: /etc/pihole/
            - name: dnsmasq
              mountPath: /etc/dnsmasq.d/
            # Mount a ram disk to /dev/shm to prevent "Bus error"
            - name: shm
              mountPath: /dev/shm
          env:
            - name: VIRTUAL_HOST
              value: "<redacted>"
            - name: DNS1
              value: "1.1.1.1"
            - name: DNS2
              value: "1.0.0.1"
            - name: WEB_PORT
              value: "31180"
          # Disabled for debugging
          #startupProbe:
          #  exec:
          #    command:
          #      - /bin/bash
          #      - -c
          #      - "dig +norecurse +retry=0 @127.0.0.1 pi.hole || exit 1"
          #  periodSeconds: 10
          #  failureThreshold: 12
          #  timeoutSeconds: 10
          #livenessProbe:
          #  exec:
          #    command:
          #      - /bin/bash
          #      - -c
          #      - "dig +norecurse +retry=0 @127.0.0.1 pi.hole || exit 1"
          #  periodSeconds: 20
          #  timeoutSeconds: 10
          imagePullPolicy: "IfNotPresent"
          resources:
            requests:
              cpu: 50m
              memory: 75Mi
            limits:
              cpu: 250m
              memory: 150Mi
          terminationMessagePath: "/dev/termination-log"
          terminationMessagePolicy: "File"
      volumes:
        - name: shm
          emptyDir:
            medium: Memory
      nodeSelector:
        kubernetes.io/hostname: tron
  volumeClaimTemplates:
    - metadata:
        name: config
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
            storage: 1.5Gi
    - metadata:
        name: dnsmasq
      spec:
        accessModes: [ "ReadWriteOnce" ]
        resources:
          requests:
            storage: 10Mi
---
kind: Service
apiVersion: v1
metadata:
  name: pihole
  namespace: home
spec:
  selector:
    app: pihole
  ports:
  - name: http
    protocol: TCP
    port: 80
    targetPort: http
```

3. any additional info to help reproduce

The "kill usage" message seems to be caused by the `pgrep pihole-FTL` command in the line below failing because the pihole-FTL command is not started (the TODO comment recognising this could be an issue).

https://github.com/pi-hole/docker-pi-hole/blob/76be304d79d8c8820575cf68128e0e6b53b57d82/s6/debian-root/etc/cont-init.d/20-start.sh#L25

There doesn't seem to be an issue causing pihole-FTL to fail as entering the container and running `service pihole-FTL start` fixes the issue and DNS queries begin resolving.

## These common fixes didn't work for my issue

- [x] I have tried removing/destroying my container, and re-creating a new container
- [ ] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [x] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [ ] I have tried running without my volume data mounts to eliminate volumes as the cause

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

**Comment by PromoFaux** [comment]
Hmm, thanks for the detailed report, though I haven't a clue where to begin trying to debug an kubernetes setup.

That said, I thought we had seen this before and indeed fixed it.

Given the following (v basic) docker-compose: (copied from README here)

```yml
version: "3"

# More info at https://github.com/pi-hole/docker-pi-hole/ and https://docs.pi-hole.net/
services:
  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "80:80/tcp"
    environment:
      TZ: 'America/Chicago'
      # WEBPASSWORD: 'set a secure password here or it will be random'
    # Volumes store your data between container upgrades
    #volumes:
    #  - './etc-pihole/:/etc/pihole/'
    #  - './etc-dnsmasq.d/:/etc/dnsmasq.d/'
    # Recommended but not required (DHCP needs NET_ADMIN)
    #   https://github.com/pi-hole/docker-pi-hole#note-on-capabilities
    cap_add:
      - NET_ADMIN
    restart: unless-stopped
```

I don't see the error:

```
pihole    | [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
pihole    | [s6-init] ensuring user provided files have correct perms...exited 0.
pihole    | [fix-attrs.d] applying ownership & permissions fixes...
pihole    | [fix-attrs.d] 01-resolver-resolv: applying...
pihole    | [fix-attrs.d] 01-resolver-resolv: exited 0.
pihole    | [fix-attrs.d] done.
pihole    | [cont-init.d] executing container initialization scripts...
pihole    | [cont-init.d] 20-start.sh: executing...
pihole    |  ::: Starting docker specific checks & setup for docker pihole/pihole
pihole    | Assigning random password: IKmnZy7y
pihole    |
pihole    |   [i] Installing configs from /etc/.pihole...
pihole    |   [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
pihole    | Existing DNS servers detected in setupVars.conf. Leaving them alone
pihole    |   [✓] New password set
pihole    | DNSMasq binding to default interface: eth0
pihole    | Added ENV to php:
pihole    |                     "PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
pihole    |                     "ServerIP" => "0.0.0.0",
pihole    |                     "VIRTUAL_HOST" => "0.0.0.0",
pihole    | Using IPv4 and IPv6
pihole    | ::: setup_blocklists now setting default blocklists up:
pihole    | ::: TIP: Use a docker volume for /etc/pihole/adlists.list if you want to customize for first boot
pihole    | ::: Blocklists (/etc/pihole/adlists.list) now set to:
pihole    | https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
pihole    | ::: Testing pihole-FTL DNS: FTL started!
pihole    | ::: Testing lighttpd config: Syntax OK
pihole    | ::: All config checks passed, cleared for startup ...
pihole    | ::: Enabling Query Logging
pihole    |   [i] Enabling logging...
  [✓] Logging has been enabled!
pihole    |  ::: Docker start setup complete
pihole    |   Pi-hole version is v5.3.1 (Latest: v5.3.1)
pihole    |   AdminLTE version is v5.5 (Latest: v5.5)
pihole    |   FTL version is v5.8.1 (Latest: v5.8.1)
pihole    | [cont-init.d] 20-start.sh: exited 0.
pihole    | [cont-init.d] done.
pihole    | [services.d] starting services
pihole    | Starting crond
pihole    | Starting pihole-FTL (no-daemon) as root
pihole    | Starting lighttpd
pihole    | [services.d] done.
```

if it's of any use:

```
 root  ~/docktest  docker images pihole/pihole:v5.8.1
REPOSITORY      TAG       IMAGE ID       CREATED        SIZE
pihole/pihole   v5.8.1    257fd58747c5   2 months ago   335MB
```

**Comment by deosrc** [comment]
Thanks @PromoFaux 

It occurs to me that even with the kill command issues resolved, Pi-Hole is still proceeding with the startup (evidenced by the version log messages). So there is a further problem which I haven't found. For some reason FTL is failing to start when the container first launches but is starting fine when I go in and start it manually.

I'm not too familiar with s6. Is there somewhere it logs failures in service startup?

In the meantime, I'm just preparing a PR to resolve the TODO comments.

**Comment by PromoFaux** [comment]
> I'm not too familiar with s6. 

That makes two of us, then. Everything I know about how this container fits together has been picked up as I go along!

I notice you are mounting /dev/shm - does the container start ok if you leave out that mount?

Also take a look at /bar/log/pihole-FTL.log to see if FTL is complaining about anything.

I've never tried it in docker, but you should be able to enable additional debugging options for FTL via it's config file:

https://docs.pi-hole.net/ftldns/configfile/#debugging-options

**Comment by deosrc** [comment]
> I notice you are mounting /dev/shm - does the container start ok if you leave out that mount?

If I remember correctly, this is a workaround for Kubernetes. It's similar to the docker `--shm` argument. Without it the container will start throwing "Bus Error" messages after a while and stop serving DNS requests.

Thanks for the info. I'll have a look and report back with what I find.

**Comment by deosrc** [comment]
Further testing required but I think I've sorted it.

FTL was complaining about a missing `CAP_SYS_NICE` permission in the logs but I don't think this was preventing it from starting, only stopping it from running optimally. I've added this and the error seems to have gone away but startup was still failing. Might want to test this and add it to the docker example.

Root cause I believe has been a classic PEBKAC error. The startup timer was set too low. From the log messages it looked like FTL had finished starting but there was still a bit more to do.

I'll close this an re-open if there's anything further to report. Thanks for your help @PromoFaux !
